Meme/Meme/NetworkClient.swift

```diff
@@ -0,0 +1,86 @@
+//
+//  NetworkClient.swift
+//  Meme
+//
+//  Created by Mesfin Bekele Mekonnen on 7/29/19.
+//  Copyright © 2019 Mesfin Bekele Mekonnen. All rights reserved.
+//
+
+import Foundation
+
+class NetworkClient {
+
+    let templateParser: TemplateParser
+    let memeParser: MemeParser
+    let templateExampleParser: TemplateExampleParser
+    var templates = [Template]()
+    var templateExample = TemplateExample(name: "", exampleURL: URL(string: "https://www.apple.com")!)
+    var meme = Meme(name: "", imageURL: URL(string: "https://www.apple.com")!)
+
+    init(templateParser: TemplateParser, memeParser: MemeParser, templateExampleParser: TemplateExampleParser) {
+        self.templateParser = templateParser
+        self.memeParser = memeParser
+        self.templateExampleParser = templateExampleParser
+    }
+
+    fileprivate func getTemplates(_ data: Data?) -> [Template]? {
+        if let data = data {
+            if let templates = self.templateParser.parse(data: data) {
+                return templates
+            }
+        }
+        return nil
+    }
+
+    fileprivate func getTemplateExample(_ data: Data?) -> TemplateExample? {
+        if let data = data {
+            if let templateExample = self.templateExampleParser.parse(data: data) {
+                return templateExample
+            }
+        }
+        return nil
+    }
+
+    fileprivate func getMemes(_ data: Data?, memeName: String) -> Meme? {
+        if let data = data {
+            if let meme = self.memeParser.parse(imageData: data, memeName: memeName) {
+                return meme
+            }
+        }
+        return nil
+    }
+
+    func requestTemplates()  {
```

Comment:
as is this is synchronous, but should probably take in a callback or return a promise

Reply:
Yep yep, I was thinking callback. 👍

Reply:
Are you familiar with promises? If not we can give it a go! Or we can use the combine framework. No pressure though

Reply:
Not familiar with promises, I think they have it in JS right. I'm all for learning. Syntax for blocks/closers in Swift looks pretty though :)

Reply:
We'll stick with closures for now and then move to futures/observables once you've mastered closures.